### PR TITLE
refactor(_DirectiveMetadataVisitor): simplify the code

### LINKED ISSUE #2493: refactor directive_metadata_reader.dart
(relates to #2430)

directive_metadata_reader.dart create an empty instance of `DirectiveMetadata` and then fill the fields as it read them from the annotation.

After #2430, there is no more a 1 to 1 mapping between annotation fields and `DirectiveMetadata` properties (`host` get split into `hostListeners`, `hostProperties`, `hostAttributes` and `hostActions`).

A better strategy would now to parse all the field from the annotation and call the new `DirectiveMetadata.create()` which takes the annotation fields.

/cc @tbosch @kegluneq

### PR DESCRIPTION
**Pull request by vicb**
fixes #2493

### CODE REVIEW

File: modules/angular2/src/transform/common/directive_metadata_reader.dart [hunk 1]
```diff
@@ -145,11 +185,12 @@ class _DirectiveMetadataVisitor extends Object
...
+    _checkMeta();
+    selector = _expressionToString(selectorValue, 'Directive#selector');
   }
 
   void _checkMeta() {
-    if (meta == null) {
+    if (!hasMeta) {
       throw new ArgumentError(
           'Incorrect value passed to readDirectiveMetadata. '
           'Expected types are Annotation and InstanceCreationExpression');
```
Comment (kegluneq) [inline modules/angular2/src/transform/common/directive_metadata_reader.dart]: Could you update this to include `NodeList` and `ListLiteral` (my omission, specified on line 11)?

File: modules/angular2/src/transform/common/directive_metadata_reader.dart [hunk 2]
```diff
@@ -48,35 +48,72 @@ num _getDirectiveType(String annotationName, Element element) {
 /// [RegisterType] object and pulling out [DirectiveMetadata].
 class _DirectiveMetadataVisitor extends Object
     with RecursiveAstVisitor<Object> {
-  DirectiveMetadata meta;
+  bool hasMeta = false;
```
Comment (kegluneq) [inline modules/angular2/src/transform/common/directive_metadata_reader.dart]: I'd suggest having this as a private getter:
```
bool get _hasMeta => type != null;
```
but I don't think it's a big deal if you prefer to leave.

File: modules/angular2/src/transform/common/directive_metadata_reader.dart [hunk 3]
```diff
@@ -48,35 +48,72 @@ num _getDirectiveType(String annotationName, Element element) {
...
+
+  // Annotation fields
+  num type;
```
Comment (kegluneq) [inline modules/angular2/src/transform/common/directive_metadata_reader.dart]: Make these all private?

**Comment by vicb** [comment]
Thanks for the feedback, all implemented. Waiting for Travis before merging.

**Comment by vicb** [comment]
landed as https://github.com/angular/angular/commit/b4e82b8bc78ef4252c562b3ab40787264efc13a5 + previous commits

**Pull request closed without merging**